server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -125,30 +134,71 @@ static void executeInternal(SearchContext searchContext) throws QueryPhaseExecut
             }
 
-            final LinkedList<QueryCollectorContext> collectors = new LinkedList<>();
-            // whether the chain contains a collector that filters documents
-            boolean hasFilterCollector = false;
+            // create the top docs collector last when the other collectors are known
+            final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(
+                searchContext,
+                searchContext.parsedPostFilter() != null || searchContext.minimumScore() != null
+            );
+
+            Collector collector = wrapWithProfilerCollectorIfNeeded(
+                searchContext.getProfilers(),
+                topDocsFactory.collector(),
+                topDocsFactory.profilerName,
+                null
+            );
+
             if (searchContext.terminateAfter() != SearchContext.DEFAULT_TERMINATE_AFTER) {
                 // add terminate_after before the filter collectors
                 // it will only be applied on documents accepted by these filter collectors
-                collectors.add(createEarlyTerminationCollectorContext(searchContext.terminateAfter()));
+                List<Collector> subCollectors = new ArrayList<>();
+                subCollectors.add(new EarlyTerminatingCollector(EMPTY_COLLECTOR, searchContext.terminateAfter(), true));
+                subCollectors.add(collector);
+                collector = wrapWithProfilerCollectorIfNeeded(
+                    searchContext.getProfilers(),
+                    MultiCollector.wrap(subCollectors),
+                    REASON_SEARCH_TERMINATE_AFTER_COUNT,
+                    collector
+                );
             }
             if (searchContext.parsedPostFilter() != null) {
                 // add post filters before aggregations
                 // it will only be applied to top hits
-                collectors.add(createFilteredCollectorContext(searcher, searchContext.parsedPostFilter().query()));
-                // this collector can filter documents during the collection
-                hasFilterCollector = true;
+                final Weight filterWeight = searcher.createWeight(
+                    searcher.rewrite(searchContext.parsedPostFilter().query()),
+                    ScoreMode.COMPLETE_NO_SCORES,
+                    1f
+                );
+                collector = wrapWithProfilerCollectorIfNeeded(
+                    searchContext.getProfilers(),
+                    new FilteredCollector(collector, filterWeight),
+                    REASON_SEARCH_POST_FILTER,
+                    collector
+                );
             }
             if (searchContext.getAggsCollector() != null) {
                 // plug in additional collectors, like aggregations
-                collectors.add(createAggsCollectorContext(searchContext.getAggsCollector()));
+                List<Collector> subCollectors = new ArrayList<>();
+                subCollectors.add(collector);
```

Comment:
I wonder if we can move this complexity in the `wrapWithProfilerCollectorIfNeeded` method. The ide awould be to use varargs in the method like:

```
 private static Collector wrapWithProfilerCollectorIfNeeded(
        Profilers profilers,
        Collector collector,
        String profilerName,
        Collector... children
    ) {
        if (profilers == null) {
            return collector;
        }
        final List<InternalProfileCollector> profileCollectors;
        if (children == null) {
            profileCollectors = List.of();
        } else {
            profileCollectors = new ArrayList<>(children.length);
            Arrays.stream(children).forEach(c -> profileCollectors.add((InternalProfileCollector) c));
        }
        return new InternalProfileCollector(
            collector,
            profilerName,
            profileCollectors
        );
    }
```

So here we can just call the method like:

```
// plug in additional collectors, like aggregations
                // in this case we pass both collectors as children profile collectors
                collector = wrapWithProfilerCollectorIfNeeded(
                    searchContext.getProfilers(),
                    MultiCollector.wrap(collector, searchContext.getAggsCollector()),
                    REASON_SEARCH_MULTI,
                    collector,
                    searchContext.getAggsCollector());
```

Reply:
I tried that, it looked a bit awkward. I will try again.

Reply:
I think this will work, #95427 will also help in that direction so that InternalProfileCollector no longer requires a list.

Reply:
The casting needed was a bit on the way and did not combine well with the varargs, but I found a way to reuse the wrap method for aggs too. It's not perfect but it's better than before I think.